op-service/eth/blob.go

```diff
@@ -70,60 +80,193 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
-			break
-		}
+
+	// first field element encodes the version and the length of the data in [1:5]
+	b[VersionOffset] = EncodingVersion
+
+	// Encode the length as big-endian uint24 into [2:5] bytes of the first field element.
+	// The length check above ensures we can always fit the length value into only 3 bytes.
+	ilen := uint32(len(data))
+	b[2] = byte(ilen >> 16)
+	b[3] = byte(ilen >> 8)
+	b[4] = byte(ilen)
+
+	// round 0 is special cased here since we can only encode the first (31-5)+1=27 bytes of input
+	// into the first field element due to version/length encoding already occupying the first 5
+	// bytes.
+	ipos := 0 // current position of the input data from where we are reading
+	buffer, ipos := read(ipos, 27, data)
+	x, ipos := read(ipos, 1, data)
+	encodedByte := x[0] & 0b0011_1111
+	b[0] = encodedByte
+	b.write(buffer, 5)
+	opos := 32 // current position of the blob where we are writing
+
+	buffer, ipos = read(ipos, 31, data)
+	y, ipos := read(ipos, 1, data)
+
+	encodedByte = (y[0] & 0b0000_1111) | ((x[0] & 0b1100_0000) >> 2)
+	opos = b.writeByteAndBuffer(opos, encodedByte, buffer)
+
+	buffer, ipos = read(ipos, 31, data)
+	z, ipos := read(ipos, 1, data)
+	encodedByte = z[0] & 0b0011_1111
+	opos = b.writeByteAndBuffer(opos, encodedByte, buffer)
+
+	buffer, ipos = read(ipos, 31, data)
+	encodedByte = ((z[0] & 0b1100_0000) >> 2) | ((y[0] & 0b1111_0000) >> 4)
+	opos = b.writeByteAndBuffer(opos, encodedByte, buffer)
+
+	for round := 1; round < Rounds && ipos < len(data); round++ {
+		buffer, ipos = read(ipos, 31, data)
+		x, ipos = read(ipos, 1, data)
+		encodedByte = x[0] & 0b0011_1111
+		opos = b.writeByteAndBuffer(opos, encodedByte, buffer)
+
+		buffer, ipos = read(ipos, 31, data)
+		y, ipos = read(ipos, 1, data)
+		encodedByte = (y[0] & 0b0000_1111) | ((x[0] & 0b1100_0000) >> 2)
+		opos = b.writeByteAndBuffer(opos, encodedByte, buffer)
+
+		buffer, ipos = read(ipos, 31, data)
+		z, ipos = read(ipos, 1, data)
+		encodedByte = z[0] & 0b0011_1111
+		opos = b.writeByteAndBuffer(opos, encodedByte, buffer)
+
+		buffer, ipos = read(ipos, 31, data)
+		encodedByte = ((z[0] & 0b1100_0000) >> 2) | ((y[0] & 0b1111_0000) >> 4)
+		opos = b.writeByteAndBuffer(opos, encodedByte, buffer)
 	}
-	if offset < len(data) {
-		return fmt.Errorf("failed to fit all data into blob. bytes remaining: %v", len(data)-offset)
+
+	if ipos < len(data) {
+		return fmt.Errorf("failed to fit all data into blob. bytes remaining: %v", len(data)-ipos)
 	}
+
 	return nil
 }
 
+// read reads up to numBytes of input from the data starting at the given offset, and returns an
+// updated offset of max(offset+numBytes, len(data)).  It always returns a byte array of length
+// exactly numBytes, even if there wasn't enough data to read. In the short-read case, any trailing
+// bytes are zero padded.
+func read(offset int, numBytes int, data []byte) ([]byte, int) {
+	byteArray := make([]byte, numBytes)
+	if offset >= len(data) {
+		return byteArray, len(data)
+	}
```

Comment:
We should not be allocating a new slice of bytes for every read call.
If we make `read` internal` to the decoding function,  then we can safely avoid the `ipos` input/output, and just return a slice of the original  bytes. This then makes it not allocate so much, and much less verbose. I'll submit a quick PR.

Reply:
https://github.com/roberto-bayardo/optimism/pull/10